### PR TITLE
[@types/lodash] Enhance lodash.get with Template Literal Types

### PR DESCRIPTION
**Pull request by lizheming**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <[<url here>](https://lodash.com/docs/4.17.15#get)>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/lodash/lodash-tests.ts [hunk 1]
```diff
@@ -5343,6 +5345,7 @@ fp.now(); // $ExpectType number
     _.chain({ a: value }).get("a", defaultValue); // $ExpectType StringChain | PrimitiveChain<false> | PrimitiveChain<true>
     _.chain({ a: undefined }).get("a", defaultValue); // $ExpectType PrimitiveChain<false> | PrimitiveChain<true>
     _.chain({ a: [1] }).get("a", []).map((val) => val.toFixed()); // $ExpectType CollectionChain<string>
+    _.chain({ a: [{ b: { c: [3] } }] }).get("a[0].b.c").map((val) => val.toFixed());  // $ExpectType CollectionChain<string>
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts:5348]: Should this be CollectionChain<number>?

**Review by aj-r — CHANGES_REQUESTED** [review]
Neat! My only concern is that this adds a lot of complexity - I'm a little worried that this could have unintended consequences although I don't imagine anything specific going wrong.

File: types/lodash/lodash-tests.ts (hunk 1 above)
Comment (aj-r) [reply types/lodash/lodash-tests.ts:5348]: Oh I see, it calls toFixed() which makes it a string

**Review by aj-r — APPROVED** [review]

**Review by sandersn — APPROVED** [review]

**Comment by lizheming** [comment]
Ready to merge

**Comment by lizheming** [comment]
BTW, most of code was created by @atsikov on the post https://dev.to/tipsy_dev/advanced-typescript-reinventing-lodash-get-4fhe, I just create the pull request. Thanks for @atsikov working for that~

**Comment by atsikov** [comment]
Thanks for putting it into PR @lizheming 🙌 
My only concern here is that these types heavily rely on literal templates, which was only added in TS 4.1, so probably a proper versioning should is needed.

**Comment by rmehner** [comment]
Thank you @lizheming,

however with this change, the following will wrongly throw an TS error:

```ts
import { get } from "lodash"

get({}, 'does.not.exist', "").trim() // will throw: Object is possibly 'undefined'.(2532)
```

Playground: https://www.typescriptlang.org/play?ssl=3&ssc=37&pln=3&pc=1#code/JYWwDg9gTgLgBAbzgcwKbwL5wGZQiOAIgBsIATAQwGcALQgWACgm0YAKBDAGjgHIyIqKgDoAdhBjDUAD2BUYvHoUIBKYTCig2KoA

We provide a default value as third argument, so it cannot be undefined.

**Comment by atsikov** [comment]
I tested this package on a large codebase, and unfortunately it's not very backwards-compatible.
Besides the above mentioned provided default, it also doesn't handle generic `string` keys, returning `undefined` instead of previous `any`.
Both issues are fixed here https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63025

**Comment by murgatroid99** [comment]
This was released as a patch but it is a breaking change for anyone using a TypeScript compiler version below 4.1. #63022 shows that this manifests as syntax errors when trying to build code that depends on these types.

**Comment by atsikov** [comment]
@murgatroid99 I checked if DefinitelyTyped allows to make a versioned override, but it looks like everything below 4.1 is considered unsupported 😕 
https://github.com/microsoft/DefinitelyTyped-tools/blob/master/packages/typescript-versions/src/index.ts#L52

As long as many packages use carets for types, minor version won't help, and major will be confusing because of lodash version mismatch.
I assume this change should be rolled back and, and `_.get` stay loosely typed forever 😞

**Comment by murgatroid99** [comment]
I did not realize that DefinitelyTyped explicitly does not support those versions of TypeScript. I can only speak for myself, but I would understand keeping the change with that justification.

**Comment by lizheming** [comment]
I saw there has an official solution for version compatibility https://github.com/DefinitelyTyped/DefinitelyTyped#i-want-to-use-features-from-very-new-typescript-versions . I'll update with that and test with Typescript <= 4.1. @atsikov May be you can realize it in your pr #63025 so we can avoid conflict?

**Comment by atsikov** [comment]
Ok, in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63025  I copied `@types/lodash@4.14.186` to `./ts4.1` which should be used for all TS versions prior to 4.2

**Comment by lizheming** [comment]
@atsikov Nice job!!! Thank you very much!!!! T_T, I'm too busy today that I have no rest time to fix it.

**Comment by lizheming** [comment]
> Thank you @lizheming,
> 
> however with this change, the following will wrongly throw an TS error:
> 
> ```ts
> import { get } from "lodash"
> 
> get({}, 'does.not.exist', "").trim() // will throw: Object is possibly 'undefined'.(2532)
> ```
> 
> Playground: https://www.typescriptlang.org/play?ssl=3&ssc=37&pln=3&pc=1#code/JYWwDg9gTgLgBAbzgcwKbwL5wGZQiOAIgBsIATAQwGcALQgWACgm0YAKBDAGjgHIyIqKgDoAdhBjDUAD2BUYvHoUIBKYTCig2KoA
> 
> We provide a default value as third argument, so it cannot be undefined.

Hello @rmehner #63025 has merged before. you can update to `@types/lodash@4.14.188` to fix it. Thank you for your report~ 😄

**Pull request merged**